arwen/contexts/output.go

```diff
@@ -285,21 +298,68 @@ func (context *outputContext) AddTxValueToAccount(address []byte, value *big.Int
 
 // GetVMOutput updates the current VMOutput and returns it
 func (context *outputContext) GetVMOutput() *vmcommon.VMOutput {
-	if context.outputState.ReturnCode == vmcommon.Ok {
-		runtime := context.host.Runtime()
-		metering := context.host.Metering()
+	runtime := context.host.Runtime()
+	metering := context.host.Metering()
 
-		gasUsed := metering.GasUsedByContract()
-		account, _ := context.GetOutputAccount(runtime.GetSCAddress())
-		account.GasUsed = math.AddUint64(account.GasUsed, gasUsed)
+	remainedFromForwarded := uint64(0)
+	account, _ := context.GetOutputAccount(runtime.GetSCAddress())
+	if context.outputState.ReturnCode == vmcommon.Ok {
+		account.GasUsed, remainedFromForwarded = metering.GasUsedByContract()
 		context.outputState.GasRemaining = metering.GasLeft()
+
+		// backward compatibility
+		if !context.host.IsArwenV2Enabled() && account.GasUsed > metering.GetGasProvided() {
+			return context.CreateVMOutputInCaseOfError(arwen.ErrNotEnoughGas)
+		}
+	} else {
+		account.GasUsed = math.AddUint64(account.GasUsed, metering.GetGasProvided())
 	}
 
 	context.removeNonUpdatedCode(context.outputState)
 
+	err := context.checkGas(remainedFromForwarded)
+	if err != nil {
+		return context.CreateVMOutputInCaseOfError(err)
+	}
+
 	return context.outputState
 }
 
+func (context *outputContext) checkGas(remainedFromForwarded uint64) error {
+	if context.host.IsArwenV2Enabled() == false {
+		return nil
+	}
+
+	gasUsed := uint64(0)
+	gasLockSentForward := false
+	for _, outputAccount := range context.outputState.OutputAccounts {
+		gasUsed = math.AddUint64(gasUsed, outputAccount.GasUsed)
+		for _, outputTransfer := range outputAccount.OutputTransfers {
+			gasUsed = math.AddUint64(gasUsed, outputTransfer.GasLimit)
+			gasUsed = math.AddUint64(gasUsed, outputTransfer.GasLocked)
+
+			if outputTransfer.GasLocked > 0 {
+				gasLockSentForward = true
+			}
+		}
+	}
+
+	gasProvided := context.host.Metering().GetGasProvided()
+	if gasLockSentForward &&
```

Comment:
Can you extract this boolean `gasLockSentForward &&...` into a variable? It's a bit hard to understand why do we add to the `gasProvided`, but a variable with a suggestive name would help

Reply:
done